DIPs/1NNN-RAC.md

```diff
@@ -0,0 +1,333 @@
+All other changes can come from this knowledge.
+
+## Prior Work
+
+In 2016 Benjamin Thaut did a talk at DConf ``D's Import and Export Business`` which covered a proposal for improving the language and the implementation details of export and shared libraries.
+
+This DIP, while not directly based upon it and does not touch upon the implementation details, does come to similar conclusions that ``export`` must not be a visibility modifier to make D work when using positive annotation.
+
+A key difference is that Benjamin's talk suggested that ``dllimport`` switch set to all was a good solution of specifying that a symbol is in DllImport mode.
+However after many years experience with D and shared libraries, it has been shown in the community usage that this is a source of constant linker errors and so a solution that is finer grained is required.
+
+## Description
+
+### Cornerstone Changes
+
+This section introduces crucial changes for future updates and ensures error-free compilation.
+
+1. External Import Path Switch: A new flag ``-extI`` is added, similar to the ``-I`` switch, marking modules as external to the currently compiling binary. This should be automated by build managers using the knowledge of dependencies being shared libraries to automatically apply it instead of ``-I``.
+2. Out-of-Binary Modules: For such modules, all non-templated scopes are interpreted to implicitly have ``extern`` attribute applied to them.
+3. Symbol Import: To import a symbol in DLL mode, use both ``export`` and ``extern``. The function body's status is irrelevant.
+
+In the following sections, the reliability of templates is considered for linking against shared libraries, the D interface generator, different export annotations and inlining.
+
+### Reliability of Templates
+
+Templates are a very easy way to make builds fail to link, due to assumed template instantiations existing.
+
+It is important therefore, that symbols originating from templates are by default not able to be exported or put into dllimport mode.
+Instead they must be reinstantiated when they are seen to be outside the binary and placed into target binary with appropriate duplication flags set.
+To do this, use the external import path switch flagging per module.
+
+Any compiler that needs to minimize the amount of code generation it performs, may do so by applying a pinning approach.
+A template instantiation is said to be pinned, iff it has been referenced by a non-templated symbol in the same module it was declared in.
+This includes variable declarations (such as globals) that are not found under a template or function parameters and return types.
+
+Once a template and its symbols has been pinned, the compiler may elide the code generation for these symbols when out of binary, and respect the exportation/dllimport symbol modes.
+
+### D Interface Generator
```

Comment:
Yes.

The D interface generator must work, right now it cannot work for both static and shared libraries.

As for template pinning yes, otherwise you would be stuck with only instantiated templates within your own binary increasing its size.